### PR TITLE
feat(multi-chain): Add support for multiple chains

### PR DESCRIPTION
**Pull request by 0xKitsune**
This PR adds support for multiple chains, allowing a client to specify which chain they would like an inclusion proof for. This PR also includes various architecture which result in faster syncing, proof generation and more efficient tree versioning. Further testing and error handling in progress.

### CODE REVIEW

**Comment by ewoolsey** [comment]
This looks killer Daniel! Will probably take me a while to get though all of these changes but I'll let you know once I'm through it!

File: src/tree/mod.rs [hunk 1]
```diff
@@ -1,103 +1,343 @@
...
 pub type PoseidonTree<Version> = LazyMerkleTree<PoseidonHash, Version>;
 pub type Hash = <PoseidonHash as Hasher>::Hash;
 
-pub const SYNC_TO_HEAD_SLEEP_SECONDS: u64 = 5;
-
-/// An abstraction over a tree with a history of changes
-///
-/// In our data model the `tree` is the oldest available tree.
-/// The entires in `tree_history` represent new additions to the tree.
 pub struct WorldTree<M: Middleware> {
```
Comment (ewoolsey) [inline src/tree/mod.rs:36]: Would be awesome to get some comments here describing how all of these pieces connect together

File: src/tree/tree_manager.rs [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+    ) -> JoinHandle<()>;
+
+    fn tree_changed_signature() -> H256;
+}
+
+pub struct TreeManager<M: Middleware, T: TreeVersion> {
+    pub address: H160,
+    pub block_scanner: Arc<BlockScanner<M>>,
+    pub chain_id: u64,
+    #[allow(dead_code)]
```
Comment (ewoolsey) [inline src/tree/tree_manager.rs:41]: How come you're not using PhantomData?

Comment (0xKitsune) [reply src/tree/tree_manager.rs]: Good catch, I meant to update to use PhantomData. I just pushed a commit with the changes.

File: src/tree/identity_tree.rs [hunk 3]
```diff
@@ -0,0 +1,592 @@
...
+            });
+
+            node_idx = (node_idx - 1) / 2;
+        }
+
+        Ok(semaphore::merkle_tree::Proof(proof))
+    }
+
+    pub fn insert(&mut self, index: u32, value: Hash) {
+        //TODO: do we want to return an error if a non unique leaf is being inserted?
```
Comment (ewoolsey) [inline src/tree/identity_tree.rs:156]: Is there any reason why we wouldn't want to return an error in that case?

File: src/tree/identity_tree.rs [hunk 4]
```diff
@@ -0,0 +1,592 @@
...
+        self.leaves.insert(value, index);
+        // We can expect here because the `reallocate` implementation for Vec<H::Hash> as DynamicTreeStorage does not fail
```
Comment (ewoolsey) [inline src/tree/identity_tree.rs:158]: Hmmm. Yeah the return type here should probably be dependent on the underlying storage. I'll look into cleaning that up on my side.

File: src/tree/block_scanner.rs [hunk 5]
```diff
@@ -1,68 +1,87 @@
...
+            let logs = result?;
+
+            aggregated_logs.extend(logs);
+        }
+
         self.last_synced_block
             .store(last_synced_block, Ordering::SeqCst);
 
-        tracing::info!(?last_synced_block, "Last synced block updated");
+        tracing::info!(chain_id = ?self.chain_id, ?last_synced_block);
```
Comment (Dzejkop) [inline src/tree/block_scanner.rs:83]: Can we add a message? Logs without messages are hard to find

File: src/tree/block_scanner.rs [hunk 6]
```diff
@@ -1,68 +1,87 @@
...
 
     /// Retrieves events matching the specified address and topics from the last synced block to the latest block, stepping by `window_size`.
+    /// Note that the logs are unsorted and should be handled accordingly.
     pub async fn next(&self) -> Result<Vec<Log>, M::Error> {
         let latest_block = self.middleware.get_block_number().await?.as_u64();
         let mut last_synced_block =
             self.last_synced_block.load(Ordering::SeqCst);
-        let mut logs = Vec::new();
 
+        let mut tasks = FuturesUnordered::new();
```
Comment (Dzejkop) [inline src/tree/block_scanner.rs:52]: Consider changing to `FuturesOrdered` to preserve ordering

File: src/abi.rs [hunk 7]
```diff
@@ -10,11 +10,6 @@ abigen!(
         function deleteIdentities(uint256[8] calldata deletionProof, uint32 batchSize, bytes calldata packedDeletionIndices, uint256 preRoot, uint256 postRoot) external
```
Comment (Dzejkop) [inline src/abi.rs:10]: Have we ever used this call on prod? If not I would rather we remove this variant and whatever env (dev/stage) it was used in we should nuke

File: src/tree/identity_tree.rs [hunk 8]
```diff
@@ -0,0 +1,612 @@
...
+            LeafUpdates::Delete(leaves) => leaves,
+        }
+    }
+}
+
+// Node index to hash, 0 indexed from the root
+pub type StorageUpdates = HashMap<u32, Hash>;
+
+// Leaf index to hash, 0 indexed from the initial leaf
+pub type Leaves = HashMap<u32, Hash>;
```
Comment (Dzejkop) [inline src/tree/identity_tree.rs:30]: Consider introducing newtypes:
```
pub struct NodeIndex(pub u32);
pub struct LeafIndex(pub u32);
```

and then discard the type aliases. That way you get type safety and fewer indirections

File: src/tree/mod.rs [hunk 9]
```diff
@@ -1,103 +1,363 @@
...
+
+        Ok(handles)
+    }
 
-            tracing::info!(?sync_time, "WorldTree synced to chain head");
-            synced.store(true, Ordering::Relaxed);
+    fn handle_tree_updates(
+        &self,
+        mut leaf_updates_rx: Receiver<(Root, LeafUpdates)>,
+        mut bridged_root_rx: Receiver<(u64, Hash)>,
```
Comment (Dzejkop) [inline src/tree/mod.rs:94]: Consider a newtype `ChainId(pub u64)`

Comment (Dzejkop) [reply src/tree/mod.rs:94]: Maybe something like this https://github.com/worldcoin/contract-deployer/blob/main/src/types.rs

File: src/tree/identity_tree.rs (hunk 8 above)
Comment (Dzejkop) [reply src/tree/identity_tree.rs:30]: Maybe something like https://github.com/worldcoin/contract-deployer/blob/main/src/types.rs

File: src/tree/mod.rs [hunk 10]
```diff
@@ -1,103 +1,363 @@
...
+    ) -> JoinHandle<()> {
+        let identity_tree = self.identity_tree.clone();
+        let chain_state = self.chain_state.clone();
+        let canonical_chain_id = self.canonical_tree_manager.chain_id;
 
+        tokio::spawn(async move {
             loop {
-                tree_updater.sync_to_head(&tree_data).await?;
+                async {
+                    tokio::select! {
```
Comment (Dzejkop) [inline src/tree/mod.rs:103]: Is there a good reason why we have to select here? Could we just spawn 2 tasks - one for canonical and the other for bridged channel proceesing?

Comment (Dzejkop) [reply src/tree/mod.rs:103]: I'm not super opposed to the idea btw. I just think maybe that way we could reduce nesting and simplify this/these loops

File: src/tree/service.rs [hunk 11]
```diff
@@ -95,92 +59,54 @@ impl<M: Middleware> TreeAvailabilityService<M> {
...
     }
 }
 
 #[derive(Serialize, Deserialize, Debug)]
 #[serde(rename_all = "camelCase", deny_unknown_fields)]
 pub struct InclusionProofRequest {
     pub identity_commitment: Hash,
-    pub root: Option<Hash>,
+    chain_id: Option<ChainId>,
 }
```
Comment (Dzejkop) [inline src/tree/service.rs:76]: Let's make both `pub` or none of them (I prefer `pub`)

File: src/tree/mod.rs (hunk 10 above)
Comment (0xKitsune) [reply src/tree/mod.rs:103]: Yeah, Ill update to spawn two tasks. This will be cleaner.

File: src/tree/service.rs [hunk 12]
```diff
@@ -95,92 +59,54 @@ impl<M: Middleware> TreeAvailabilityService<M> {
...
+        Ok(handles)
     }
 }
 
 #[derive(Serialize, Deserialize, Debug)]
 #[serde(rename_all = "camelCase", deny_unknown_fields)]
 pub struct InclusionProofRequest {
     pub identity_commitment: Hash,
-    pub root: Option<Hash>,
+    chain_id: Option<ChainId>,
```
Comment (Dzejkop) [inline src/tree/service.rs:75]: Add `#[serde(skip_serializing_if = "Option::is_none")]` otherwise we'll always serialize this as `"chainId":null`

Comment (Dzejkop) [inline src/tree/service.rs:75]:
```suggestion
    #[serde(skip_serializing_if = "Option::is_none")]
```

File: src/tree/service.rs [hunk 13]
```diff
@@ -95,92 +59,54 @@ impl<M: Middleware> TreeAvailabilityService<M> {
...
-    }
-}
-
-#[derive(Debug, Serialize, Deserialize)]
-#[serde(rename_all = "camelCase")]
-pub struct SyncResponse {
-    pub synced: bool,
-    pub block_number: Option<u64>,
-}
+    let chain_id = query_params.chain_id.or(req.chain_id);
```
Comment (Dzejkop) [inline src/tree/service.rs:102]: It's kinda weird that query params chain id overrides the one from the request. Maybe let's ditch the request one?

File: src/tree/identity_tree.rs [hunk 14]
```diff
@@ -0,0 +1,612 @@
...
+
+impl PartialOrd for Root {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+pub struct IdentityTree {
+    pub tree: DynamicMerkleTree<PoseidonHash>,
+    pub tree_updates: BTreeMap<Root, StorageUpdates>,
```
Comment (Dzejkop) [inline src/tree/identity_tree.rs:71]: Not necessary in this PR but let's remember to add metrics that display stats about these various hash maps (how many entries we have in them). I feel like it'll be easy to forget to remove entries

Comment (0xKitsune) [reply src/tree/identity_tree.rs:71]: Sounds good, I am planning to open a PR to update logging/metrics after this is merged.

File: src/tree/identity_tree.rs [hunk 15]
```diff
@@ -0,0 +1,611 @@
...
+                for (idx, val) in updates.iter() {
+                    self.leaves.insert(*val, *idx);
+                }
+            }
+            LeafUpdates::Delete(ref updates) => {
+                for (_, val) in updates.iter() {
+                    self.leaves.remove(val);
+                }
+            }
+        }
```
Comment (Dzejkop) [inline src/tree/identity_tree.rs:187]: Not sure why I feel this way, but `ref` feels like a leftover from the days before match ergonomics were a thing.

```suggestion
        match &leaf_updates {
            LeafUpdates::Insert(updates) => {
                for (idx, val) in updates.iter() {
                    self.leaves.insert(*val, *idx);
                }
            }
            LeafUpdates::Delete(updates) => {
                for (_, val) in updates.iter() {
                    self.leaves.remove(val);
                }
            }
        }
```

File: src/tree/identity_tree.rs [hunk 16]
```diff
@@ -0,0 +1,610 @@
...
+            //TODO: Use a more efficient approach than to clone the last update
+            update.1.clone()
+        } else {
+            HashMap::new()
+        };
+
+        while let Some(node_idx) = node_queue.pop_back() {
+            // Check if the parent is already in the updates hashmap, indicating it has already been calculated
+            //TODO: note why we set to 0 if idx is 0
+            let parent_idx = if node_idx == 0 { 0 } else { (node_idx - 1) / 2 };
```
Comment (Dzejkop) [inline src/tree/identity_tree.rs:211]: Replacing `0` with `continue` here and skipping that one iteration seems to work here (tests in this file pass)

File: src/tree/identity_tree.rs [hunk 17]
```diff
@@ -0,0 +1,610 @@
...
+            if updates.contains_key(&parent_idx.into()) {
+                continue;
+            }
+
+            let left_sibling_idx = node_idx * 2 + 1;
+            let right_sibling_idx = node_idx * 2 + 2;
```
Comment (Dzejkop) [inline src/tree/identity_tree.rs:217]: These are child node idxs, no?

File: src/tree/mod.rs [hunk 18]
```diff
@@ -1,103 +1,430 @@
...
+                .await?
+                .as_u64();
+
+            bridged_tree
+                .block_scanner
+                .last_synced_block
+                .store(current_block, Ordering::SeqCst);
+        }
+
+        // Get all logs from the canonical tree from the last synced block to the chain tip
```
Comment (Dzejkop) [inline src/tree/mod.rs:207]: This actually gets all logs from that last synced to `min(current block, last synced + window size)`. So this function won't actually sync to head

File: src/tree/identity_tree.rs [hunk 19]
```diff
@@ -0,0 +1,652 @@
...
+
+        // Traverse the tree from the leaf to the root, constructing the proof along the way with precedence for the updated node values
+        while node_idx > 0 {
+            let sibling_idx = if node_idx % 2 == 0 {
+                node_idx - 1
+            } else {
+                node_idx + 1
+            };
+
+            // Check if the sibling is in the updates, otherwise get the node from the tree
```
Comment (Dzejkop) [inline src/tree/identity_tree.rs:272]: If `updates` is per a given `root` is there a change that we'll miss the node updates from the "previous root".

Let's say we have 0x03 on mainnet, 0x02 on polygon and 0x01 on optimism.
So in this case we'd have the dense memory tree on 0x01 and then 0x02 and 0x03 in `tree_updates`, right?

We fetch an inclusion proof using root 0x03, but a node that was updated in 0x02 is in the path of the inclusion proof. This code however only checks the updates for 0x03 so we miss that update from 0x02 and return an invalid inclusion proof.

Not sure if I'm reading this right, but we should definitely have a test case for this

Comment (Dzejkop) [reply src/tree/identity_tree.rs:272]: Ok, I was wrong here, I see now that `construct_storage_updates` flattens the previous updates.

https://github.com/worldcoin/world-tree/pull/71

Added a test

**Review by Dzejkop — APPROVED** [review]

**Pull request merged**